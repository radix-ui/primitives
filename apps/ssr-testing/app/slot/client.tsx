'use client';

import * as React from 'react';
import { Slot } from 'radix-ui';

export const Link = React.forwardRef<
  React.ComponentRef<'a'>,
  React.ComponentProps<'a'> & { asChild?: boolean }
>(({ asChild = false, ...props }, forwardedRef) => {
  const Comp = asChild ? Slot.Root : 'a';
  return <Comp {...props} ref={forwardedRef} />;
});

export const LinkSlottable = React.forwardRef<
  React.ComponentRef<'a'>,
  React.ComponentProps<'a'> & { asChild?: boolean }
>(({ asChild = false, ...props }, forwardedRef) => {
  const Comp = asChild ? Slot.Root : 'a';
  return (
    <Comp {...props} ref={forwardedRef}>
      <span>left</span>
      <Slot.Slottable>{props.children}</Slot.Slottable>
      <span>right</span>
    </Comp>
  );
});

export const LinkButton = React.forwardRef<
  React.ComponentRef<typeof Link>,
  React.ComponentProps<typeof Link>
>((props, forwardedRef) => (
  <Button asChild>
    <Link {...props} ref={forwardedRef}>
      {props.children}
    </Link>
  </Button>
));

export const Button = React.forwardRef<
  React.ComponentRef<'button'>,
  React.ComponentProps<'button'> & { asChild?: boolean }
>(({ asChild = false, ...props }, forwardedRef) => {
  const Comp = asChild ? Slot.Root : 'button';
  return <Comp {...props} ref={forwardedRef} style={{ display: 'flex', gap: '3rem' }} />;
});

export const ButtonSlottable = React.forwardRef<
  React.ComponentRef<'button'>,
  React.ComponentProps<'button'> & { asChild?: boolean }
>(({ children, asChild = false, ...props }, forwardedRef) => {
  const Comp = asChild ? Slot.Root : 'button';
  return (
    <Comp {...props} ref={forwardedRef} style={{ display: 'flex', gap: '3rem' }}>
      <span>left</span>
      <Slot.Slottable>{children}</Slot.Slottable>
      <span>right</span>
    </Comp>
  );
});

export const ButtonNestedSlottable = React.forwardRef<
  React.ComponentRef<typeof Button>,
  React.ComponentProps<typeof Button>
>(({ children, asChild = false, ...props }, forwardedRef) => {
  const Comp = asChild ? Slot.Root : 'button';
  return (
    <Comp {...props} ref={forwardedRef} style={{ display: 'flex', gap: '3rem' }}>
      <Slot.Slottable child={children}>
        {(slottable) => (
          <>
            <span>left</span>
            <b>bold {slottable}</b>
            <span>right</span>
          </>
        )}
      </Slot.Slottable>
    </Comp>
  );
});

export const IconButtonNestedSlottable = React.forwardRef<
  React.ComponentRef<typeof Button>,
  React.ComponentProps<typeof Button>
>(({ children, ...props }, forwardedRef) => {
  return (
    <Button {...props} ref={forwardedRef} style={{ display: 'flex', gap: '3rem' }}>
      <Slot.Root>
        <Slot.Slottable child={children}>
          {(slottable) => (
            <>
              <span>ICON</span>
              <b>bold {slottable}</b>
            </>
          )}
        </Slot.Slottable>
      </Slot.Root>
    </Button>
  );
});
