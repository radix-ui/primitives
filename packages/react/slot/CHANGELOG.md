# @radix-ui/react-slot

## 1.2.3

- Replace deprecated 'ElementRef' with 'ComponentRef' (#3426)
- Revert slot changes causing errors in server components

## 1.2.2

- Add `use client` to slot entry to prevent RSC errors

## 1.2.1

- Use stable composed refs in SlotClone ([#3477](https://github.com/radix-ui/primitives/pull/3477))
