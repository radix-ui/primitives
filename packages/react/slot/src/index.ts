export {
  Slot,
  Slottable,
  //
  Root,
  createSlot,
  createSlottable,
} from './slot';
export type { SlotProps } from './slot';
