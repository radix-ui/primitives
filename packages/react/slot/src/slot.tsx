import * as React from 'react';
import { composeRefs } from '@radix-ui/react-compose-refs';

/* -------------------------------------------------------------------------------------------------
 * Slot
 * -----------------------------------------------------------------------------------------------*/

interface SlotProps extends React.HTMLAttributes<HTMLElement> {
  children?: React.ReactNode;
}

/* @__NO_SIDE_EFFECTS__ */ export function createSlot(ownerName: string) {
  const SlotClone = createSlotClone(ownerName);
  const Slot = React.forwardRef<HTMLElement, SlotProps>((props, forwardedRef) => {
    const { children, ...slotProps } = props;
    const childrenArray = React.Children.toArray(children);
    const slottable = childrenArray.find(isSlottable);

    if (slottable) {
      // the new element to render is the one passed as a child of `Slottable`
      const newElement = slottable.props.children;

      const newChildren = childrenArray.map((child) => {
        if (child === slottable) {
          // because the new element will be the one rendered, we are only interested
          // in grabbing its children (`newElement.props.children`)
          if (React.Children.count(newElement) > 1) return React.Children.only(null);
          return React.isValidElement(newElement)
            ? (newElement.props as { children: React.ReactNode }).children
            : null;
        } else {
          return child;
        }
      });

      return (
        <SlotClone {...slotProps} ref={forwardedRef}>
          {React.isValidElement(newElement)
            ? React.cloneElement(newElement, undefined, newChildren)
            : null}
        </SlotClone>
      );
    }

    return (
      <SlotClone {...slotProps} ref={forwardedRef}>
        {children}
      </SlotClone>
    );
  });

  Slot.displayName = `${ownerName}.Slot`;
  return Slot;
}

const Slot = createSlot('Slot');

/* -------------------------------------------------------------------------------------------------
 * SlotClone
 * -----------------------------------------------------------------------------------------------*/

interface SlotCloneProps {
  children: React.ReactNode;
}

/* @__NO_SIDE_EFFECTS__ */ function createSlotClone(ownerName: string) {
  const SlotClone = React.forwardRef<any, SlotCloneProps>((props, forwardedRef) => {
    const { children, ...slotProps } = props;

    if (React.isValidElement(children)) {
      const childrenRef = getElementRef(children);
      const props = mergeProps(slotProps, children.props as AnyProps);
      // do not pass ref to React.Fragment for React 19 compatibility
      if (children.type !== React.Fragment) {
        props.ref = forwardedRef ? composeRefs(forwardedRef, childrenRef) : childrenRef;
      }
      return React.cloneElement(children, props);
    }

    return React.Children.count(children) > 1 ? React.Children.only(null) : null;
  });

  SlotClone.displayName = `${ownerName}.SlotClone`;
  return SlotClone;
}

/* -------------------------------------------------------------------------------------------------
 * Slottable
 * -----------------------------------------------------------------------------------------------*/

const SLOTTABLE_IDENTIFIER = Symbol('radix.slottable');

interface SlottableProps {
  children: React.ReactNode;
}

interface SlottableComponent extends React.FC<SlottableProps> {
  __radixId: symbol;
}

/* @__NO_SIDE_EFFECTS__ */ export function createSlottable(ownerName: string) {
  const Slottable: SlottableComponent = ({ children }) => {
    return <>{children}</>;
  };
  Slottable.displayName = `${ownerName}.Slottable`;
  Slottable.__radixId = SLOTTABLE_IDENTIFIER;
  return Slottable;
}

const Slottable = createSlottable('Slottable');

/* ---------------------------------------------------------------------------------------------- */

type AnyProps = Record<string, any>;

function isSlottable(
  child: React.ReactNode
): child is React.ReactElement<SlottableProps, typeof Slottable> {
  return (
    React.isValidElement(child) &&
    typeof child.type === 'function' &&
    '__radixId' in child.type &&
    child.type.__radixId === SLOTTABLE_IDENTIFIER
  );
}

function mergeProps(slotProps: AnyProps, childProps: AnyProps) {
  // all child props should override
  const overrideProps = { ...childProps };

  for (const propName in childProps) {
    const slotPropValue = slotProps[propName];
    const childPropValue = childProps[propName];

    const isHandler = /^on[A-Z]/.test(propName);
    if (isHandler) {
      // if the handler exists on both, we compose them
      if (slotPropValue && childPropValue) {
        overrideProps[propName] = (...args: unknown[]) => {
          const result = childPropValue(...args);
          slotPropValue(...args);
          return result;
        };
      }
      // but if it exists only on the slot, we use only this one
      else if (slotPropValue) {
        overrideProps[propName] = slotPropValue;
      }
    }
    // if it's `style`, we merge them
    else if (propName === 'style') {
      overrideProps[propName] = { ...slotPropValue, ...childPropValue };
    } else if (propName === 'className') {
      overrideProps[propName] = [slotPropValue, childPropValue].filter(Boolean).join(' ');
    }
  }

  return { ...slotProps, ...overrideProps };
}

// Before React 19 accessing `element.props.ref` will throw a warning and suggest using `element.ref`
// After React 19 accessing `element.ref` does the opposite.
// https://github.com/facebook/react/pull/28348
//
// Access the ref using the method that doesn't yield a warning.
function getElementRef(element: React.ReactElement) {
  // React <=18 in DEV
  let getter = Object.getOwnPropertyDescriptor(element.props, 'ref')?.get;
  let mayWarn = getter && 'isReactWarning' in getter && getter.isReactWarning;
  if (mayWarn) {
    return (element as any).ref;
  }

  // React 19 in DEV
  getter = Object.getOwnPropertyDescriptor(element, 'ref')?.get;
  mayWarn = getter && 'isReactWarning' in getter && getter.isReactWarning;
  if (mayWarn) {
    return (element.props as { ref?: React.Ref<unknown> }).ref;
  }

  // Not DEV
  return (element.props as { ref?: React.Ref<unknown> }).ref || (element as any).ref;
}

export {
  Slot,
  Slottable,
  //
  Slot as Root,
};
export type { SlotProps };
