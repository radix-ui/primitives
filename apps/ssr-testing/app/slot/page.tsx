import * as React from 'react';
import * as Client from './client';
import * as Server from './server';

export default function Page() {
  return (
    <>
      <p>All components should be rendered as links</p>

      <h2>Client.LinkButton</h2>

      <Client.LinkButton href="/">children</Client.LinkButton>

      <h2>Client.Button as Client.Link</h2>

      <Client.Button asChild>
        <Client.Link href="/">children</Client.Link>
      </Client.Button>

      <h2>Client.Button as Server.Link</h2>

      <Client.Button asChild>
        <Server.Link href="/">children</Server.Link>
      </Client.Button>

      <h2>Client.Button as Client.LinkSlottable</h2>

      <Client.Button asChild>
        <Client.LinkSlottable href="/">children</Client.LinkSlottable>
      </Client.Button>

      <h2>Client.Button as Server.LinkSlottable</h2>

      <Client.Button asChild>
        <Server.LinkSlottable href="/">children</Server.LinkSlottable>
      </Client.Button>

      <h2>Client.ButtonSlottable as Server.Link</h2>

      <Client.ButtonSlottable asChild>
        <Server.Link href="/">children</Server.Link>
      </Client.ButtonSlottable>

      <h2>Client.ButtonSlottable as Client.Link</h2>

      <Client.ButtonSlottable asChild>
        <Client.Link href="/">children</Client.Link>
      </Client.ButtonSlottable>

      <h2>Client.ButtonNestedSlottable as Server.Link</h2>

      <Client.ButtonNestedSlottable asChild>
        <Server.Link href="/">children</Server.Link>
      </Client.ButtonNestedSlottable>

      <h2>Client.ButtonNestedSlottable as Client.Link</h2>

      <Client.ButtonNestedSlottable asChild>
        <Client.Link href="/">children</Client.Link>
      </Client.ButtonNestedSlottable>

      <h2>Client.IconButtonNestedSlottable as Server.Link</h2>

      <Client.IconButtonNestedSlottable asChild>
        <Server.Link href="/">children</Server.Link>
      </Client.IconButtonNestedSlottable>

      <h2>Client.IconButtonNestedSlottable as Client.Link</h2>

      <Client.IconButtonNestedSlottable asChild>
        <Client.Link href="/">children</Client.Link>
      </Client.IconButtonNestedSlottable>

      <hr />

      <h2>Server.LinkButton</h2>

      <Server.LinkButton href="/">children</Server.LinkButton>

      <h2>Server.Button as Server.Link</h2>

      <Server.Button asChild>
        <Server.Link href="/">children</Server.Link>
      </Server.Button>

      <h2>Server.Button as Client.Link</h2>

      <Server.Button asChild>
        <Client.Link href="/">children</Client.Link>
      </Server.Button>

      <h2>Server.Button as Server.LinkSlottable</h2>

      <Server.Button asChild>
        <Server.LinkSlottable href="/">children</Server.LinkSlottable>
      </Server.Button>

      <h2>Server.Button as Client.LinkSlottable</h2>

      <Server.Button asChild>
        <Client.LinkSlottable href="/">children</Client.LinkSlottable>
      </Server.Button>

      <h2>Server.ButtonSlottable as Client.Link</h2>

      <Server.ButtonSlottable asChild>
        <Client.Link href="/">children</Client.Link>
      </Server.ButtonSlottable>

      <h2>Server.ButtonSlottable as Server.Link</h2>

      <Server.ButtonSlottable asChild>
        <Server.Link href="/">children</Server.Link>
      </Server.ButtonSlottable>

      <h2>Server.ButtonNestedSlottable as Client.Link</h2>

      <Server.ButtonNestedSlottable asChild>
        <Client.Link href="/">children</Client.Link>
      </Server.ButtonNestedSlottable>

      <h2>Server.ButtonNestedSlottable as Server.Link</h2>

      <Server.ButtonNestedSlottable asChild>
        <Server.Link href="/">children</Server.Link>
      </Server.ButtonNestedSlottable>

      <h2>Server.IconButtonNestedSlottable as Server.Link</h2>

      <Server.IconButtonNestedSlottable asChild>
        <Server.Link href="/">children</Server.Link>
      </Server.IconButtonNestedSlottable>

      <h2>Server.IconButtonNestedSlottable as Client.Link</h2>

      <Server.IconButtonNestedSlottable asChild>
        <Client.Link href="/">children</Client.Link>
      </Server.IconButtonNestedSlottable>
    </>
  );
}
