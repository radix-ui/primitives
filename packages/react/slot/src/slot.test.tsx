import * as React from 'react';
import { cleanup, render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Slot, Slottable } from './slot';
import { afterEach, describe, it, beforeEach, vi, expect } from 'vitest';

describe('given a slotted Trigger', () => {
  afterEach(cleanup);
  describe('with onClick on itself', () => {
    const handleClick = vi.fn();

    beforeEach(() => {
      handleClick.mockReset();
      render(
        <Trigger as={Slot} onClick={handleClick}>
          <button type="button">Click me</button>
        </Trigger>
      );
      fireEvent.click(screen.getByRole('button'));
    });

    it('should call the onClick passed to the Trigger', async () => {
      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('with onClick on the child', () => {
    const handleClick = vi.fn();

    beforeEach(() => {
      handleClick.mockReset();
      render(
        <Trigger as={Slot}>
          <button type="button" onClick={handleClick}>
            Click me
          </button>
        </Trigger>
      );
      fireEvent.click(screen.getByRole('button'));
    });

    it("should call the child's onClick", async () => {
      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('with onClick on itself AND the child', () => {
    const handleTriggerClick = vi.fn();
    const handleChildClick = vi.fn();

    beforeEach(() => {
      handleTriggerClick.mockReset();
      handleChildClick.mockReset();
      render(
        <Trigger as={Slot} onClick={handleTriggerClick}>
          <button type="button" onClick={handleChildClick}>
            Click me
          </button>
        </Trigger>
      );
      fireEvent.click(screen.getByRole('button'));
    });

    it("should call the Trigger's onClick", async () => {
      expect(handleTriggerClick).toHaveBeenCalledTimes(1);
    });

    it("should call the child's onClick", async () => {
      expect(handleChildClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('with onClick on itself AND undefined onClick on the child', () => {
    const handleTriggerClick = vi.fn();

    beforeEach(() => {
      handleTriggerClick.mockReset();
      render(
        <Trigger as={Slot} onClick={handleTriggerClick}>
          <button type="button" onClick={undefined}>
            Click me
          </button>
        </Trigger>
      );
      fireEvent.click(screen.getByRole('button'));
    });

    it("should call the Trigger's onClick", async () => {
      expect(handleTriggerClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('with undefined onClick on itself AND onClick on the child', () => {
    const handleChildClick = vi.fn();

    beforeEach(() => {
      handleChildClick.mockReset();
      render(
        <Trigger as={Slot} onClick={undefined}>
          <button type="button" onClick={handleChildClick}>
            Click me
          </button>
        </Trigger>
      );
      fireEvent.click(screen.getByRole('button'));
    });

    it("should call the child's onClick", async () => {
      expect(handleChildClick).toHaveBeenCalledTimes(1);
    });
  });
});

describe('given a Button with Slottable', () => {
  afterEach(cleanup);
  describe('without asChild', () => {
    it('should render a button with icon on the left/right', async () => {
      const tree = render(
        <Button iconLeft={<span>left</span>} iconRight={<span>right</span>}>
          Button <em>text</em>
        </Button>
      );

      expect(tree.container).toMatchSnapshot();
    });
  });

  describe('with asChild', () => {
    it('should render a link with icon on the left/right', async () => {
      const tree = render(
        <Button iconLeft={<span>left</span>} iconRight={<span>right</span>} asChild>
          <a href="https://radix-ui.com">
            Button <em>text</em>
          </a>
        </Button>
      );

      expect(tree.container).toMatchSnapshot();
    });
  });
});

// TODO: Unskip when underlying issue is resolved
// Reverted in https://github.com/radix-ui/primitives/pull/3554
describe.skip('given an Input', () => {
  const handleRef = vi.fn();

  beforeEach(() => {
    handleRef.mockReset();
  });

  afterEach(cleanup);

  describe('without asChild', () => {
    it('should only call function refs once', async () => {
      render(<Input ref={handleRef} />);
      await userEvent.type(screen.getByRole('textbox'), 'foo');
      expect(handleRef).toHaveBeenCalledTimes(1);
    });
  });

  describe('with asChild', () => {
    it('should only call function refs once', async () => {
      render(
        <Input asChild ref={handleRef}>
          <input />
        </Input>
      );
      await userEvent.type(screen.getByRole('textbox'), 'foo');
      expect(handleRef).toHaveBeenCalledTimes(1);
    });
  });
});

type TriggerProps = React.ComponentProps<'button'> & { as: React.ElementType };

const Trigger = ({ as: Comp = 'button', ...props }: TriggerProps) => <Comp {...props} />;

const Button = React.forwardRef<
  React.ComponentRef<'button'>,
  React.ComponentProps<'button'> & {
    asChild?: boolean;
    iconLeft?: React.ReactNode;
    iconRight?: React.ReactNode;
  }
>(({ children, asChild = false, iconLeft, iconRight, ...props }, forwardedRef) => {
  const Comp = asChild ? Slot : 'button';
  return (
    <Comp {...props} ref={forwardedRef}>
      {iconLeft}
      <Slottable>{children}</Slottable>
      {iconRight}
    </Comp>
  );
});

const Input = React.forwardRef<
  React.ElementRef<'input'>,
  React.ComponentProps<'input'> & {
    asChild?: boolean;
  }
>(({ asChild, children, ...props }, forwardedRef) => {
  const Comp = asChild ? Slot : 'input';
  const [value, setValue] = React.useState('');

  return (
    <Comp
      {...props}
      onChange={(event) => setValue(event.target.value)}
      ref={forwardedRef}
      value={value}
    >
      {children}
    </Comp>
  );
});
