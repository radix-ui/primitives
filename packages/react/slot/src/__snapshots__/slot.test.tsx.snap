// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`given a Button with Slottable > with asChild > should render a link with icon on the left/right 1`] = `
<div>
  <a
    href="https://radix-ui.com"
  >
    <span>
      left
    </span>
    Button 
    <em>
      text
    </em>
    <span>
      right
    </span>
  </a>
</div>
`;

exports[`given a Button with Slottable > without asChild > should render a button with icon on the left/right 1`] = `
<div>
  <button>
    <span>
      left
    </span>
    Button 
    <em>
      text
    </em>
    <span>
      right
    </span>
  </button>
</div>
`;

exports[`given a Slot with React lazy components > with a lazy component in Button with Slottable > should render a lazy link with icon on the left/right 1`] = `
<div>
  <a
    href="https://radix-ui.com"
  >
    <span>
      left
    </span>
    Button 
    <em>
      text
    </em>
    <span>
      right
    </span>
  </a>
</div>
`;
